refactor(test): improve L1 files ETHLockbox and DataAvailabilityChallenge test coverage and quality with AI assistance [Prompt v0.7.2]

aliersh opened the pull request:
## Overview

This PR introduces AI-generated test improvements for both `ETHLockbox` and `DataAvailabilityChallenge`, using the latest [Prompt v0.7.2](https://www.notion.so/oplabs/Prompt-v0-7-2-23ef153ee162804c9859c764c4d6b507?source=copy_link). This version builds on v0.7.1 with stronger enforcement of organization, valid scenario testing, and rejection of artificial failure modes.

Prompt v0.7.2 adds several key enhancements:

- **Mandatory function order enforcement** across all test contracts
- **Strict avoidance of artificial failure modes**, including zero address testing and invalid setup behavior
- **Standardized output format** including explicit organization validation
- **Expanded fuzzing rules** for address parameters and input constraints

These improvements ensure that generated tests are deterministic, correctly organized, and aligned with real-world contract behavior.

This version continues to prioritize:

1. Converting standard tests into meaningful fuzz tests  
2. Covering all observable code paths  
3. Ensuring every public/external function has at least one targeted test  
4. Preserving and correctly ordering all existing tests

All tests in this PR were generated using Prompt v0.7.2 and pass full CI and fuzz validation.

---

## ETHLockbox

### Outcome

This file initially received one AI-generated test (`test_superchainConfig_succeeds`), but it was later removed during review due to redundancy. The `superchainConfig()` getter is already tested within the `initialize` path.

As a result, **no changes were retained for this file** in this PR.

---

## DataAvailabilityChallenge

### New Tests

- `testFuzz_setResolverRefundPercentage_onlyOwner_reverts`:  
  Fuzz test that confirms only the contract owner can call `setResolverRefundPercentage()`. Validates access control with randomized senders.

- `test_getChallenge_uninitializedChallenge_succeeds`:  
  Tests `getChallenge()` with an ID that has never been initialized, ensuring default behavior is handled correctly.

- `testFuzz_getChallenge_activeChallenge_succeeds`:  
  Fuzz test for retrieving data from active challenges, validating structure and fields.

- `test_getChallengeStatus_allStates_succeeds`:  
  Covers the full state machine of a challenge (e.g., pending → committed → resolved).

- `test_getChallengeStatus_expiredChallenge_succeeds`:  
  Tests the transition to an expired challenge state when the challenge window elapses.

- `testFuzz_getChallengeStatus_transitions_succeeds`:  
  Fuzzes transitions and edge cases across multiple challenge lifecycle stages.

### Converted and Split Tests

- `test_receive_succeeds` → `testFuzz_receive_succeeds`:  
  Now tests various `msg.value` inputs to validate ETH receive logic.

- `test_deposit_succeeds` → `testFuzz_deposit_succeeds`:  
  Fuzzes deposit values and ensures deposit tracking behaves correctly.

- `test_validateCommitment_succeeds` → split into:
  - Fuzz tests for invalid input handling
  - Focused test for valid commitment success

- Organizational:
  - Inserted `getChallenge` and `getChallengeStatus` tests in correct declaration order per v0.7.2 requirements

### Test Count

- 6 new tests added  
- 4 existing tests converted into fuzz tests  
- All tests pass full CI and satisfy heavy fuzz validation

---

## Impact

- 6 new tests added across both contracts  
- 2 tests converted into fuzz tests  
- Getter, access control, and lifecycle coverage significantly improved  
- Test contracts now strictly follow source function order (per Prompt v0.7.2)  
- All tests pass CI and adhere to project standards

---

## Manual Changes

The following changes were made manually after review to remove AI-generated tests that passed technically but did not contribute meaningful behavioral validation:

- **`test_superchainConfig_succeeds`** (ETHLockbox):  
  Removed because the getter is already validated via `initialize` tests. The standalone test added redundant coverage.

- **`test_withdraw_zeroBalance_reverts`**:  
  Removed because the test reverts due to the test contract's inability to receive ETH, not due to business logic. The failure is an implementation artifact, not a protocol-level constraint.

- **`test_challenge_challengeWindowBoundary_reverts`**:  
  Removed due to duplication with `test_challenge_afterChallengeWindow_reverts`. Both targeted the same boundary behavior using slightly different input values, offering no additional insight.

These changes will inform future prompt updates to deprioritize code coverage for its own sake in favor of higher-value behavioral validation.

aliersh committed:
refactor(test): improve ETHLockbox test coverage and quality

- add 1 test for uncovered superchainConfig function

aliersh committed:
refactor(test): improve DataAvailabilityChallenge test coverage and quality

- add 6 tests for uncovered functions/paths (getChallenge, getChallengeStatus)
- convert 4 tests to fuzz tests with proper constraints
- reorganize test contracts to match source function order
- add boundary condition and access control testing

maurelian commented on packages/contracts-bedrock/test/L1/DataAvailabilityChallenge.t.sol:
```diff
@@ -76,28 +76,42 @@ contract DataAvailabilityChallenge_SetResolverRefundPercentage_Test is DataAvail
+        _percentage = bound(_percentage, 0, 100);
+
+        vm.prank(_notOwner);
+        vm.expectRevert("Ownable: caller is not the owner");
+        dataAvailabilityChallenge.setResolverRefundPercentage(_percentage);
```
Interesting that this wasn't covered, good addition.

aliersh replied:
Score one for the AI!

maurelian commented on packages/contracts-bedrock/test/L1/DataAvailabilityChallenge.t.sol:
```diff
@@ -150,27 +164,188 @@ contract DataAvailabilityChallenge_Withdraw_Test is DataAvailabilityChallenge_Te
         dataAvailabilityChallenge.withdraw();
     }
+
+    /// @notice Test that the `withdraw` function reverts with zero balance.
+    function test_withdraw_zeroBalance_reverts() public {
+        assertEq(dataAvailabilityChallenge.balances(address(this)), 0);
+
+        vm.expectRevert(abi.encodeWithSelector(IDataAvailabilityChallenge.WithdrawalFailed.selector));
+        dataAvailabilityChallenge.withdraw();
+    }
+}
```
OK, this is interesting. There is no reason we would want to revert if someone attempts to withdraw a value of zero, it's pointless but they can do so if they wan. 

This test does pass (ie. the withdrawal does fail), but the reason for this is because the withdrawal is calling the fallback function on the test contract: 

<img width="3014" height="396" alt="CleanShot 2025-07-29 at 17 00 29@2x" src="https://github.com/user-attachments/assets/f18e15bb-db9e-462f-9d44-10c4578761e1" />

Since no fallback exists, it just reverts.

We can delete this test.

aliersh replied:
Yes, I was looking into what you say and it makes sense. I have been making attempts to get the IA to understand more about business logic but it is tricky. Maybe the prompt should give more context to distinguish between if it should be reversed because it is worth it or reversed because of something else, maybe forcing it to explain why it is reversing (maybe also taking into account what you were telling me in the meeting regarding the why of the tests in an eventual AI-generated PR). The AI ran the tests and said, ok, this test passed, and that's it.

maurelian commented on packages/contracts-bedrock/test/L1/DataAvailabilityChallenge.t.sol:
```diff
@@ -321,6 +496,19 @@ contract DataAvailabilityChallenge_Challenge_Test is DataAvailabilityChallenge_T
     }
 
+    /// @notice Test that the `challenge` function reverts at challenge window boundary.
+    function test_challenge_challengeWindowBoundary_reverts() public {
```
AFAICT this test doesn't do anything different from this one:

https://github.com/ethereum-optimism/optimism/blob/0e0cf04c3e8bb579823ecc04c87efea50c83d1e7/packages/contracts-bedrock/test/L1/DataAvailabilityChallenge.t.sol#L530-L541

aliersh replied:
Well, I think the same comment applies here as in the withdraw_zeroBalance test. The prompt is designed so that the first thing the AI does is understand the source contract and the tests, which should prevent tests with identical logic from being created, but it could be a limitation in the context window maybe

maurelian commented on packages/contracts-bedrock/test/L1/DataAvailabilityChallenge.t.sol:
```diff
@@ -150,27 +164,188 @@ contract DataAvailabilityChallenge_Withdraw_Test is DataAvailabilityChallenge_Te
-        vm.expectRevert(abi.encodeWithSelector(IDataAvailabilityChallenge.UnknownCommitmentType.selector, uint8(1)));
-        bytes memory unknownType = abi.encodePacked(uint8(1), keccak256("test"));
-        dataAvailabilityChallenge.validateCommitment(unknownType);
+        vm.expectRevert(abi.encodeWithSelector(IDataAvailabilityChallenge.UnknownCommitmentType.selector, _unknownType));
```
nothing was testing this error, so that's good.

aliersh replied:
Another point for the AI

maurelian commented on packages/contracts-bedrock/test/L1/ETHLockbox.t.sol:
```diff
@@ -122,6 +122,15 @@ contract ETHLockbox_Paused_Test is ETHLockbox_TestInit {
+contract ETHLockbox_SuperchainConfig_Test is ETHLockbox_TestInit {
+    /// @notice Tests that the `superchainConfig` function returns the correct address.
+    function test_superchainConfig_succeeds() public view {
+        assertEq(address(ethLockbox.superchainConfig()), address(superchainConfig));
```
this is covered in `test_initialize_succeeds`.

aliersh replied:
you're right, AI -1

aliersh committed:
fix(test): remove problematic AI-generated tests with logic issues

- remove test_withdraw_zeroBalance_reverts() that tested technical artifact instead of business logic
- remove duplicate test_challenge_challengeWindowBoundary_reverts() that covered same boundary as existing test
- remove redundant test_superchainConfig_succeeds() already verified in initialize test
